Configurable UTC/local-time

waltonseymour opened the pull request:
This PR allows for `Plottable.Formatters.time` to display dates in UTC in addition to local time with the addition of an optional boolean `useUTC`.

This flag is also passed down from the `TimeAxis` constructor which also contains the additional `useUTC` flag.

For us, this solves a business requirement of displaying time in UTC without having to modify the underlying Date objects, which has a number of drawbacks.

waltonseymour committed:
use utc time

waltonseymour committed:
Merge pull request #1 from nautiluslabsco/utc

default UTC time

waltonseymour committed:
Run dist on install (#2)

* adds prepublish

* change to prepare

waltonseymour committed:
make utc configurable

waltonseymour committed:
fix hard-coded false

waltonseymour committed:
adds testing

themadcreator commented on src/core/formatters.ts:
```diff
@@ -244,10 +244,14 @@ export function multiTime() {
  * @returns {Formatter} A formatter for time/date values.
  */
-export function time(specifier: string): Formatter {
+export function time(specifier: string, useUTC?: boolean): Formatter {
```
Please add the default value for this optional argument: `useUTC = DEFAULT_USE_UTC` and above, `const DEFAULT_USE_UTC = false;` (also update comment to indicate default it to not use utc)

themadcreator commented on package.json:
```diff
@@ -11,6 +11,7 @@
   "main": "build/src/index.js",
   "typings": "build/src/index.d.ts",
   "scripts": {
+    "prepare": "npm run dist",
```
I'm not sure we want this. Even if we do, this isn't the PR for it, so please remove.

waltonseymour replied:
removed

themadcreator commented:
Other than my comments, this changes looks good to me. Thanks!

waltonseymour committed:
remove prepare

waltonseymour committed:
address comments

waltonseymour committed:
use utc for tick intervals

waltonseymour commented:
I've updated the PR to address comments.
I've also corrected an issue with `tickInterval` where the ticks would be generated with local time while the labels would reflect UTC time.

waltonseymour committed:
add missing parameter

themadcreator commented on src/axes/timeAxis.ts:
```diff
@@ -80,115 +80,120 @@ export class Time extends Axis<Date> {
+    ];
+  }
+
+  private _useUTC: boolean = false;
```
I believe this default will actually be overridden with `undefined` if the constructor is invoked without the third parameter.

when the object instance is first created, `this._useUTC` is initialized with `false`, but then the constructor runs, which explicitly overrides `this._useUTC` with the optional argument. You can fix this by simply moving this default to the constructor signature like `(..., useUTC: boolean = false)`. Or better yet, `(..., useUTC: boolean = DEFAULT_USE_UTC)` and put `const DEFAULT_USE_UTC = false;` at the top of this file.

themadcreator replied:
Though, since you're already defining the default at a lower level, you could just not initialize `this._useUTC`

waltonseymour replied:
Given that default values are set at `time` and `tickInterval`, I think it's fine to leave it potentially undefined. I've removed the initialized value for `_useUTC`

themadcreator reviewed (changes requested):
Really close, just one minor thing

waltonseymour committed:
remove initialized value
